miasm2/expression/simplifications_common.py

```diff
@@ -1075,13 +1099,101 @@ def simp_cond_eq_zero(expr_s, expr):
     new_expr = ExprCond(arg1, expr.src2, expr.src1)
     return new_expr
 
+def simp_sign_inf_zeroext(expr_s, expr):
+    """
+    X.zeroExt() <=s 0 => X == 0
+    X.zeroExt() <=s cst => X <=u cst[:X.size] (cond)
```

Comment:
First, this equation is not true if `X.zeroExt` is the same size of `X`, cause it would be rewrote as `X <=s cst => X <=u cst` (but it seems to be already checked by the code).

Second, maybe we can put it with a kind of more general rule:
X.zeroExt(size) with size > X.size is always positive.

A positive element <s or <=s X has three possibilities. Either X is always positive, and the 's' can be converted to a 'u' (so `X.zeroExt() <=s cst` => `X.zeroExt() <=u cst` if `cst` is positive, or `X.zeroExt <=s Y.zeroExt` => `X.zeroExt <=u Y.zeroExt`); either X is always negative, and this become always false or a comparison to 0; either we don't know and it can't be simplified.

With this rule, we can then add the rule `X.zeroExt() <=u cst` => `X <=u cst[:X.size]`, and `X <=u 0` => `X == 0`.

Those rules seem wider, more general, to me. What do you think?

Note: re-reading myself, I actually wonder if `X.zeroExt() <=s cst => X <=u cst[:X.size]` is even actually correct for negative `cst`. Like with X = 5, cst = 0xFFFFFFFF, 0x00000005 is not <=s 0xFFFFFFFF, but 0x5 is <=u 0xF

Reply:
```
 I actually wonder if X.zeroExt() <=s cst => X <=u cst[:X.size] is even actually correct for negative cst
```
Not in all cases: if bits are stripped of during slice, it can be false:
```
ExprInt(0xFF, 8).zeroExt(16) <=s 0x7FF
```
cannot be sliced.


Anyway, I am re reading your comment to try to understand the final goal of the code.

Reply:
Sorry but I dont understand your comment:
```
Those rules seem wider, more general, to me. What do you think?
```
What is/are your final reduction rule(s)?

Don't forget that  `simp_test_zeroext_inf` has additional conditions:
```
X.zeroExt() <=u cst => X <=u cst[:X.size]
```
Like previously, not valid in each case.

Reply:
We can add this rule:
`X.zeroExt() <=s cst` =>
* if cst is positive, `X.zeroExt() <=u cst`. Justification: signed comparison of positive number is the same than unsigned comparison
* if cst is strictly negative, it's always False. Justification: left part is a positive or null number, right part is strictly negative

This simplification is quite simple, and will report on the `simp_test_zeroext_inf` simplification the slice extraction / special case handling. In addition, we obtain the `False` case (when cst is strictly negative).

Reply:
Done